### PR TITLE
Write guide about how to use the catalog to query for all packages

### PR DESCRIPTION
**Pull request by joelverhagen**
Address https://github.com/NuGet/NuGetGallery/issues/4914.

I picked this scenario by looking at the highest cost (sum total duration) V2 queries done today (top 7):

1. Find reverse dependencies
1. Get most recently updated packages
1. `GetUpdates()`
1. Page over all latest stable
1. `Search()`
1. Substrings on metadata fields
1. Page over all packages

image

Items 1, 2, 6, 7 are all easily solved using the catalog. Reverse dependencies may one day have 1st class support. So I thought a generic guide on how to use the catalog was the best first guide to write for the NuGet V3 API.

Review URL:
https://review.docs.microsoft.com/en-us/nuget/guides/api/query-for-all-published-packages?branch=jver-v2-v3-guides

/cc @emgarten @xavierdecoster @ryuyu @scottbommarito

### CODE REVIEW

**Comment by saldana** [comment]
### :white_check_mark: Validation status: passed


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[docs/API/catalog-resource.md](https://github.com/NuGet/docs.microsoft.com-nuget/blob/jver-v2-v3-guides/docs/API/catalog-resource.md) | :white_check_mark:Succeeded | |
[docs/Guides/api/query-for-all-published-packages.md](https://github.com/NuGet/docs.microsoft.com-nuget/blob/jver-v2-v3-guides/docs/Guides/api/query-for-all-published-packages.md) | :white_check_mark:Succeeded | |
[docs/TOC.md](https://github.com/NuGet/docs.microsoft.com-nuget/blob/jver-v2-v3-guides/docs/TOC.md) | :white_check_mark:Succeeded | |

For more details, please refer to the [build report](https://opbuildstorageprod.blob.core.windows.net/report/2017%5C11%5C3%5C19a7edc0-f91d-0fd5-6e5f-4d131db05c11%5CPullRequest%5C201711032027353691-553%5Cworkflow_report.html?sv=2015-02-21&sr=b&sig=LB5sqkAeazMYSNQzPSv7pWrdcBT%2F7iTwyhzl7I8M0rc%3D&st=2017-11-03T20%3A24%3A10Z&se=2017-12-04T20%3A29%3A10Z&sp=r).

**Note:** If you changed an existing file name or deleted a file, broken links in other files to the deleted or renamed file are listed only in the full build report.

**Comment by joelverhagen** [comment]
@kraigb, please wait to merge this until I give a go ahead. This one needs some dev/PM eyes.

**Comment by kraigb** [comment]
OK, I created a do-not-merge label that you can use for anything else like this.

File: docs/Guides/api/query-for-all-published-packages.md [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+catalog, see its [API reference document](../../api/catalog-resource.md).
+
+The following steps can be implemented in any programming language of your choice. If you want a full running sample,
+take a look at the [C# sample](#c-sample-code) mentioned below.
+
+Otherwise, follow the guide below to build a reliable catalog reader.
+
+## Initialize a cursor
+
+The first piece of a reliable catalog reader is implementing a cursor. For full details about the design of a catalog
```
Comment (skofman1) [inline docs/Guides/api/query-for-all-published-packages.md]: Perhaps: "The first step in building.."

File: docs/Guides/api/query-for-all-published-packages.md [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+cursor, see the [catalog reference document](../../api/catalog-resource.md#cursor). In short, cursor is just a point
```
Comment (skofman1) [inline docs/Guides/api/query-for-all-published-packages.md]: Remove "just"

File: docs/Guides/api/query-for-all-published-packages.md [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+
+For this guide, we'll initialize our cursor to a timestamp one hour ago. For now, just save that timestamp in memory.
+
+```cs
+DateTime cursor = DateTime.UtcNow.AddHours(-1);
+```
+
+## Determine catalog index URL
+
+The location of every resource (endpoint) in the NuGet API should be should be discovered using the
```
Comment (skofman1) [inline docs/Guides/api/query-for-all-published-packages.md]: should be should be (twice)

File: docs/Guides/api/query-for-all-published-packages.md (hunk 1 above)
Comment (joelverhagen) [reply docs/Guides/api/query-for-all-published-packages.md]: Fixed.

File: docs/Guides/api/query-for-all-published-packages.md (hunk 2 above)
Comment (joelverhagen) [reply docs/Guides/api/query-for-all-published-packages.md]: Fixed.

File: docs/Guides/api/query-for-all-published-packages.md (hunk 3 above)
Comment (joelverhagen) [reply docs/Guides/api/query-for-all-published-packages.md]: Fixed.

**Comment by saldana** [comment]
### :white_check_mark: Validation status: passed


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[docs/API/catalog-resource.md](https://github.com/NuGet/docs.microsoft.com-nuget/blob/jver-v2-v3-guides/docs/API/catalog-resource.md) | :white_check_mark:Succeeded | |
[docs/Guides/api/query-for-all-published-packages.md](https://github.com/NuGet/docs.microsoft.com-nuget/blob/jver-v2-v3-guides/docs/Guides/api/query-for-all-published-packages.md) | :white_check_mark:Succeeded | |
[docs/TOC.md](https://github.com/NuGet/docs.microsoft.com-nuget/blob/jver-v2-v3-guides/docs/TOC.md) | :white_check_mark:Succeeded | |

For more details, please refer to the [build report](https://opbuildstorageprod.blob.core.windows.net/report/2017%5C11%5C3%5C19a7edc0-f91d-0fd5-6e5f-4d131db05c11%5CPullRequest%5C201711032129108276-553%5Cworkflow_report.html?sv=2015-02-21&sr=b&sig=79Cd%2FlZfEfBaIdpXXQBAA06H2%2BRmUPE62C83oPELZiA%3D&st=2017-11-03T21%3A25%3A10Z&se=2017-12-04T21%3A30%3A10Z&sp=r).

**Note:** If you changed an existing file name or deleted a file, broken links in other files to the deleted or renamed file are listed only in the full build report.

**Comment by joelverhagen** [comment]
@kraigb, this is ready to merge given your sign-off.

**Comment by saldana** [comment]
### :white_check_mark: Validation status: passed


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[docs/API/catalog-resource.md](https://github.com/NuGet/docs.microsoft.com-nuget/blob/jver-v2-v3-guides/docs/API/catalog-resource.md) | :white_check_mark:Succeeded | |
[docs/Guides/api/query-for-all-published-packages.md](https://github.com/NuGet/docs.microsoft.com-nuget/blob/jver-v2-v3-guides/docs/Guides/api/query-for-all-published-packages.md) | :white_check_mark:Succeeded | |
[docs/TOC.md](https://github.com/NuGet/docs.microsoft.com-nuget/blob/jver-v2-v3-guides/docs/TOC.md) | :white_check_mark:Succeeded | |

For more details, please refer to the [build report](https://opbuildstorageprod.blob.core.windows.net/report/2017%5C11%5C7%5C19a7edc0-f91d-0fd5-6e5f-4d131db05c11%5CPullRequest%5C201711071750442949-553%5Cworkflow_report.html?sv=2015-02-21&sr=b&sig=gRyHb5lpru9DAe1nO7v9rareGBhH9eBSaE4lngJmnfk%3D&st=2017-11-07T17%3A46%3A42Z&se=2017-12-08T17%3A51%3A42Z&sp=r).

**Note:** If you changed an existing file name or deleted a file, broken links in other files to the deleted or renamed file are listed only in the full build report.

**Pull request merged**